MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,2046 @@
+
+    template <typename T>
+      void print_track_set(const T& ts) const {
+        for (auto r : ts)
+          printf(" %u", r.key());
+      }
+
+    template <typename T>
+      void print_track_set(const T& ts, const reco::Vertex& v) const {
+        for (auto r : ts)
+          printf(" %u%s", r.key(), (v.trackWeight(r) < mfv::track_vertex_weight_min ? "!" : ""));
+      }
+
+    void print_track_set(const reco::Vertex& v) const {
+      for (auto r = v.tracks_begin(), re = v.tracks_end(); r != re; ++r)
+        printf(" %lu%s", r->key(), (v.trackWeight(*r) < mfv::track_vertex_weight_min ? "!" : ""));
+    }
+
+    bool is_track_subset(const track_set& a, const track_set& b) const {
+      bool is_subset = true;
+      const track_set& smaller = a.size() <= b.size() ? a : b;
+      const track_set& bigger  = a.size() <= b.size() ? b : a;
+
+      for (auto t : smaller)
+        if (bigger.count(t) < 1) {
+          is_subset = false;
+          break;
+        }
+
+      return is_subset;
+    }
+
+    track_set vertex_track_set(const reco::Vertex& v, const double min_weight = mfv::track_vertex_weight_min) const {
+      track_set result;
+
+      for (auto it = v.tracks_begin(), ite = v.tracks_end(); it != ite; ++it) {
+        const double w = v.trackWeight(*it);
+        const bool use = w >= min_weight;
+        assert(use);
+        //if (verbose) ("trk #%2i pt %6.3f eta %6.3f phi %6.3f dxy %6.3f dz %6.3f w %5.3f  use? %i\n", int(it-v.tracks_begin()), (*it)->pt(), (*it)->eta(), (*it)->phi(), (*it)->dxy(), (*it)->dz(), w, use);
+        if (use)
+          result.insert(it->castTo<reco::TrackRef>());
+      }
+
+      return result;
+    }
+
+    track_vec vertex_track_vec(const reco::Vertex& v, const double min_weight = mfv::track_vertex_weight_min) const {
+      track_set s = vertex_track_set(v, min_weight);
+      return track_vec(s.begin(), s.end());
+    }
+
+    Measurement1D vertex_dist(const reco::Vertex& v0, const reco::Vertex& v1) const {
+      if (use_2d_vertex_dist)
+        return vertex_dist_2d.distance(v0, v1);
+      else
+        return vertex_dist_3d.distance(v0, v1);
+    }
+
+    Measurement1D miss_dist(const reco::Vertex& sv, const AlgebraicVector3& ref, const AlgebraicVector3& mom) {
+      // miss distance is magnitude of (jet direction (= n) cross (tv - sv) ( = d))
+      // to calculate uncertainty, use |n X d|^2 = (|n||d|)^2 - (n . d)^2
+      AlgebraicVector3 n = ROOT::Math::Unit(mom);
+      AlgebraicVector3 d(sv.x() - ref(0),
+          sv.y() - ref(1),
+          sv.z() - ref(2));
+      AlgebraicVector3 n_cross_d = ROOT::Math::Cross(n, d);
+      double n_dot_d = ROOT::Math::Dot(n, d);
+      double val = ROOT::Math::Mag(n_cross_d);
+      AlgebraicVector3 jac(2 * d(0) - 2 * n_dot_d * n(0),
+          2 * d(1) - 2 * n_dot_d * n(1),
+          2 * d(2) - 2 * n_dot_d * n(2));
+      return Measurement1D(val, sqrt(ROOT::Math::Similarity(jac, sv.covariance())) / 1 / val); // modified err from 2->1 of sv

```

Comment:
ahh, I think the fix can be easier than that. miss_dist actually can be replaceable by track_dist and it is only used for plotting three histograms that are counterparts of https://github.com/DisplacedVertices/cmssw-usercode/blob/0dcb7022d40fd368fe39dae00b241c0e1e8fedd9/MFVNeutralino/plugins/Vertexer.cc#L1093-L1096 

So, it was my ignorant back then that I used two different variables to compare before and after. And to remind you on this 2->1 thing, it was a modification of miss_dist similar to the one in VertexAuxProducer.cc that computed vtx-to-vtx with two sv uncertainties but for vtx-to-trk here it only needs one uncertainty so the division is no longer two.

Reply:
Oh, this rings a bell... Did we find that the two variables gave the exact same result? (including the uncertainty?)

In any case, yes if we want to simplify and use the other variable instead, let's do so!
(and perhaps add a comment about it being equivalent to miss_dist, if true)

Reply:
I just ran a quick check and the two variables can be very close in values but they are not the same. Here is a snippet of how close they are:

`tkvtxdist: 0.000395554 =? 0.0063212
tkvtxdist err: 0.00211053 =? 0.00381192
tkvtxdist sig: 0.187419 =? 1.65827
tkvtxdist: 0.00476385 =? 0.00605866
tkvtxdist err: 0.00764036 =? 0.00630391
tkvtxdist sig: 0.623511 =? 0.961095
tkvtxdist: 0.0012259 =? 0.0024107
tkvtxdist err: 0.00472256 =? 0.00628551                                                                                                                                             tkvtxdist sig: 0.259584 =? 0.383533
tkvtxdist: 0.000845326 =? 0.000900955
tkvtxdist err: 0.00330124 =? 0.00440006
tkvtxdist sig: 0.256063 =? 0.20476
tkvtxdist: 0.00387419 =? 0.00420693
tkvtxdist err: 0.00567559 =? 0.00448893
tkvtxdist sig: 0.682606 =? 0.937177
tkvtxdist: 0.00575438 =? 0.00613873
tkvtxdist err: 0.00958615 =? 0.00879883
tkvtxdist sig: 0.600281 =? 0.697675`

However, I also realize that the variable that I placed as a cut value if from tk_vtx_dist not miss_dist. I was just the histograms that got filled by miss_dist
https://github.com/DisplacedVertices/cmssw-usercode/blob/3f7bde4a4997884ccdf2f99efef94180976feac6/MFVNeutralino/plugins/Vertexer.cc#L991

Reply:
I just did a check, they're close in a few cases (within 15%), but off by up to a factor of 16 in others...

```
tkvtxdist:     0.000395554 =? 0.0063212  , ratio is 0.0625757767512
tkvtxdist err: 0.00211053 =? 0.00381192  , ratio is 0.553665869168
tkvtxdist sig: 0.187419 =? 1.65827       , ratio is 0.113020798784

tkvtxdist:     0.00476385 =? 0.00605866  , ratio is 0.786287726989
tkvtxdist err: 0.00764036 =? 0.00630391  , ratio is 1.21200334396
tkvtxdist sig: 0.623511 =? 0.961095      , ratio is 0.648750643797

tkvtxdist:     0.0012259 =? 0.0024107    , ratio is 0.50852449496
tkvtxdist err: 0.00472256 =? 0.00628551  , ratio is 0.751340782212
tkvtxdist sig: 0.259584 =? 0.383533      , ratio is 0.676823115612

tkvtxdist:     0.000845326 =? 0.000900955, ratio is 0.938255517756
tkvtxdist err: 0.00330124 =? 0.00440006  , ratio is 0.750271587206
tkvtxdist sig: 0.256063 =? 0.20476       , ratio is 1.2505518656

tkvtxdist:     0.00387419 =? 0.00420693  , ratio is 0.92090669443
tkvtxdist err: 0.00567559 =? 0.00448893  , ratio is 1.26435252944
tkvtxdist sig: 0.682606 =? 0.937177      , ratio is 0.728364012348

tkvtxdist:     0.00575438 =? 0.00613873  , ratio is 0.937389329715
tkvtxdist err: 0.00958615 =? 0.00879883  , ratio is 1.0894800786
tkvtxdist sig: 0.600281 =? 0.697675      , ratio is 0.860402049665
```

I'm surprised since I thought the actual distances were quite similar from our old checks of the code, even if the uncertainties weren't... It would be interesting to better understand this since we often use miss_dist as a debugging tool and for studies, so it would be super useful to (eventually) understand which definition we should use. For now, we should just converge on something for the purpose of being able to merge the code for the time being. We could always open a new issue targeted at studying why these differ.

Reply:
Sure. I have changed the miss_dist to trk_vtx_dist for now. You can open a new issue.